lib/streamaggr/streamaggr.go

```diff
@@ -708,83 +726,101 @@ func (a *aggregator) runFlusher(pushFunc PushFunc, alignFlushToInterval, skipInc
 		}
 	}
 
-	if dedupInterval <= 0 {
-		alignedSleep(interval)
-		t := time.NewTicker(interval)
-		defer t.Stop()
+	flushDeadline := time.Now().Truncate(interval).Add(interval)
+	tickInterval := time.Duration(a.tickInterval) * time.Millisecond
+	alignedSleep(tickInterval)
+
+	var dedupIdx, flushIdx int
 
-		if alignFlushToInterval && skipIncompleteFlush {
-			a.flush(nil, interval, true)
-			ignoreFirstIntervals--
+	t := time.NewTicker(tickInterval)
+	defer t.Stop()
+
+	isSkippedFirstFlush := false
+	for tickerWait(t) {
+		ct := time.Now()
+
+		dedupTime := ct.Truncate(tickInterval)
+		if a.ignoreOldSamples {
+			dedupIdx, flushIdx = a.getFlushIndices(dedupInterval, interval, dedupTime, flushDeadline)
 		}
+		pf := pushFunc
+
+		// Calculate delay
+		a.muFlushAfter.Lock()
+		flushAfterMsec := a.flushAfter.Quantile(0.95)
+		a.flushAfter.Reset()
+		a.muFlushAfter.Unlock()
+		flushAfter := time.Duration(flushAfterMsec) * time.Millisecond
+
+		if flushAfter > tickInterval {
+			logger.Warnf("metrics ingestion lag (%v) is more than tick interval (%v). "+
+				"gaps are expected in aggregations", flushAfter, tickInterval)
+			pf = nil
+		} else {
+			time.Sleep(flushAfter)
+		}
```

Comment:
I'm not sure if I understand the purpose of this delay.
If I understand this correctly, the idea is to postpone flush until the p95 of non-rejected samples so that there will be more samples included. However, I'm not sure if it is a good idea to enable such adaptive delay by default. It might be counterintuitive as samples will come with delay.
Could you elaborate on the use-case when this can be helpful? Maybe I just misunderstood the purpose of this part.

Reply:
window index is calculated based on a timestamp. In case, when there's a significant delay, without this change samples should be either discarded or pushed to the next window, which actually makes it useless to implement windowing at all